### PR TITLE
Creator of speed/chained markets from pending

### PR DESCRIPTION
**Pull request by lotinasa**
[TIP-202](https://github.com/thales-markets/thales-improvement-proposals/blob/main/TIPs/TIP-202.md)

1. Add Creator contract for Speed and Chained markets which stores pending markets and has method for creating speed markets from pending, including method for atomic creation of market without creating pending one first
2. AMM contracts are refactored to have only one create method (merged diff and default collateral) and moved Pyth price logic to Creator contract
3. Minor fix on SpeedMarketsAMMData for couple of old Chained markets where length of array was invalid
4. Add Pyth price publish time for strike price on Speed Market ( PR #458 )

Release instructions:

- [x] 1. Create new NPM package version which is used for zkSync contracts
- [x] 2. Deploy SpeedMarketMatercopy, SpeedMarketsAMMCreator (zkSync: SpeedMarketsCreator)
- [x] 3. Upgrade SpeedMarketsAMM (zkSync: SpeedMarkets), ChainedSpeedMarketsAMM and SpeedMarketsAMMData
- [x] 4. Config of SpeedMarketsAMMCreator: setMaxCreationDelay(30), setAddressManager, addToWhitelist(bot)
- [x] 6. Config of AddressManager: setAddressInAddressBook(SpeedMarketsAMMCreator, address), setAddressInAddressBook(ChainedSpeedMarketsAMM, address), setAddresses(_stakingThales: zero_address)
- [x] 7. Config of SpeedMarketsAMM: setAMMAddresses, setLimitParams (decrease maximumPriceDelay and maximumPriceDelayForResolving to possible minimum e.g. 15,10, decrease minBuyinAmount by 40% to 3, increase maxBuyinAmount by 5% to 52.5), setSafeBoxAndMaxSkewImpact (skewSlippage)
- [x] 8. Config of ChainedSpeedMarketsAMM: setLimitParams (decrease minBuyinAmount by 40% to 3)
- [x] 9. Transfer owner of Creator to pDAO

### CODE REVIEW

File: contracts/SpeedMarkets/SpeedMarketsAMMData.sol [hunk 1]
```diff
@@ -161,7 +161,9 @@ contract SpeedMarketsAMMData is Initializable, ProxyOwned, ProxyPausable {
             for (uint j = 0; j < market.numOfDirections(); j++) {
                 marketDirections[j] = market.directions(j);
                 if (j < market.numOfPrices()) {
-                    marketStrikePrices[j] = market.strikePrices(j);
+                    try market.strikePrices(j) {
+                        marketStrikePrices[j] = market.strikePrices(j);
+                    } catch (bytes memory) {} // fix for old Chained markets where length of prices array is invalid
```
Comment (grujawork) [inline contracts/SpeedMarkets/SpeedMarketsAMMData.sol:166]: empty catch?

**Review by grujawork — APPROVED** [review]

File: contracts/SpeedMarkets/SpeedMarketsAMMData.sol (hunk 1 above)
Comment (lotinasa) [reply contracts/SpeedMarkets/SpeedMarketsAMMData.sol:166]: Here I wanted just to skip fetching strike price from array which doesn't exist. At the beginning of release of Chained markets there was a bug on Market where length of strike prices from market itself was eg. 6 but strikePrices real length is less than 6. This was fixed in the past with new Mastercopy, however fetching data from couple of old chained markets is failing. Here just preventing failure for history data on 2 or 3 chained markets.

File: contracts/SpeedMarkets/SpeedMarkets.sol [hunk 2]
```diff
@@ -97,6 +93,44 @@ contract SpeedMarkets is Initializable, ProxyOwned, ProxyPausable, ProxyReentran
...
+        bytes32 asset;
+        uint64 strikeTime;
+        PythStructs.Price pythPrice;
+        Direction direction;
+        uint buyinAmount;
+        bool transferSusd;
+        address referrer;
+        uint skewImpact;
+        IAddressManager.Addresses contractsAddresses;
+        uint8 creatorID;
```
Comment (kirilaa) [inline contracts/SpeedMarkets/SpeedMarkets.sol:131]: what was creator ID ?

Comment (kirilaa) [reply contracts/SpeedMarkets/SpeedMarkets.sol:131]: Ok, we talked about it, to prevent market duplicates

**Review by kirilaa — DISMISSED** [review]

**Review by danijelthales — APPROVED** [review]

**Review by vladanthales — APPROVED** [review]

**Pull request merged**